File: app/src/main/java/com/hangyeollee/go4lunch/ui/main_home_activity/workmates_fragment/WorkmatesViewModel.java
package com.hangyeollee.go4lunch.ui.main_home_activity.workmates_fragment;

import android.app.Application;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

import com.hangyeollee.go4lunch.R;
import com.hangyeollee.go4lunch.data.model.LunchRestaurant;
import com.hangyeollee.go4lunch.data.model.User;
import com.hangyeollee.go4lunch.data.repository.AutoCompleteDataRepository;
import com.hangyeollee.go4lunch.data.repository.FirebaseRepository;

import java.util.ArrayList;
import java.util.List;

public class WorkmatesViewModel extends ViewModel {

    private final Application context;

    private final MediatorLiveData<WorkmatesViewState> workMatesFragmentViewStateMediatorLiveData = new MediatorLiveData<>();

    public WorkmatesViewModel(
        Application context,
        FirebaseRepository firebaseRepository,
        AutoCompleteDataRepository autoCompleteDataRepository
    ) {
        this.context = context;

        LiveData<List<User>> userListLiveData = firebaseRepository.getUsersList();
        LiveData<List<LunchRestaurant>> lunchRestaurantListLiveData = firebaseRepository.getLunchRestaurantListOfAllUsers();
        LiveData<String> userInputLiveData = autoCompleteDataRepository.getUserInputMutableLiveData();

        workMatesFragmentViewStateMediatorLiveData.addSource(userListLiveData, userList ->
            combine(userList, lunchRestaurantListLiveData.getValue(), userInputLiveData.getValue())
        );

        workMatesFragmentViewStateMediatorLiveData.addSource(lunchRestaurantListLiveData, lunchRestaurantList ->
            combine(userListLiveData.getValue(), lunchRestaurantList, userInputLiveData.getValue())
        );

        workMatesFragmentViewStateMediatorLiveData.addSource(userInputLiveData, userInput ->
            combine(userListLiveData.getValue(), lunchRestaurantListLiveData.getValue(), userInput)
        );
    }

    private void combine(@Nullable List<User> userList, @Nullable List<LunchRestaurant> lunchRestaurantList, @Nullable String userInput) {
        if (userList == null || lunchRestaurantList == null) {
            return;
        }

        List<WorkmatesItemViewState> recyclerViewItemViewStateList = new ArrayList<>();

        List<String> userIdsEatingToday = new ArrayList<>();

        for (LunchRestaurant lunchRestaurant : lunchRestaurantList) {
            User matchingUser = null;

            for (User user : userList) {
                if (user.getId().equalsIgnoreCase(lunchRestaurant.getUserId())) {
                    matchingUser = user;
                    break;
                }
            }

            if (matchingUser != null) {
                userIdsEatingToday.add(matchingUser.getId());

                if (userInput == null || matchingUser.getName().contains(userInput) || lunchRestaurant.getRestaurantName().contains(userInput)) {
                    recyclerViewItemViewStateList.add(
                        new WorkmatesItemViewState(
                            matchingUser.getPhotoUrl(),
                            matchingUser.getName(),
                            lunchRestaurant.getRestaurantName(),
                            lunchRestaurant.getRestaurantId()
                        )
                    );
                }
            }
        }

        for (User user : userList) {
            if (!userIdsEatingToday.contains(user.getId())) {
                if (userInput == null || user.getName().contains(userInput)) {
                    recyclerViewItemViewStateList.add(
                        new WorkmatesItemViewState(
                            user.getPhotoUrl(),
                            user.getName(),
                            context.getString(R.string.not_going_to_restaurant_yet),
                            null
                        )
                    );
                }
            }
        }

        workMatesFragmentViewStateMediatorLiveData.setValue(new WorkmatesViewState(recyclerViewItemViewStateList));
    }

    public LiveData<WorkmatesViewState> getWorkmatesFragmentViewStateLiveData() {
        return workMatesFragmentViewStateMediatorLiveData;
    }

}

File: app/src/main/java/com/hangyeollee/go4lunch/ui/workmates/WorkmatesViewModel.kt
package com.hangyeollee.go4lunch.ui.workmates

import android.app.Application
import com.hangyeollee.go4lunch.data.repository.FirebaseRepository
import com.hangyeollee.go4lunch.data.repository.AutoCompleteDataRepository
import androidx.lifecycle.ViewModel
import androidx.lifecycle.MediatorLiveData
import com.hangyeollee.go4lunch.ui.workmates.WorkmatesViewState
import androidx.lifecycle.LiveData
import com.hangyeollee.go4lunch.data.model.LunchRestaurant
import com.hangyeollee.go4lunch.ui.workmates.WorkmatesItemViewState
import com.hangyeollee.go4lunch.R
import com.hangyeollee.go4lunch.data.model.User
import java.util.ArrayList

class WorkmatesViewModel(
    private val context: Application,
    firebaseRepository: FirebaseRepository,
    autoCompleteDataRepository: AutoCompleteDataRepository
) : ViewModel() {
    private val workMatesFragmentViewStateMediatorLiveData = MediatorLiveData<WorkmatesViewState>()

    init {
        val userListLiveData = firebaseRepository.usersList
        val lunchRestaurantListLiveData = firebaseRepository.lunchRestaurantListOfAllUsers
        val userInputLiveData = autoCompleteDataRepository.userInputMutableLiveData

        workMatesFragmentViewStateMediatorLiveData.addSource(userListLiveData) { userList: List<User>? ->
            combine(userList, lunchRestaurantListLiveData.value, userInputLiveData.value)
        }
        workMatesFragmentViewStateMediatorLiveData.addSource(lunchRestaurantListLiveData) { lunchRestaurantList: List<LunchRestaurant>? ->
            combine(userListLiveData.value, lunchRestaurantList, userInputLiveData.value)
        }
        workMatesFragmentViewStateMediatorLiveData.addSource(userInputLiveData) { userInput: String? ->
            combine(userListLiveData.value, lunchRestaurantListLiveData.value, userInput)
        }
    }

    private fun combine(userList: List<User>?, lunchRestaurantList: List<LunchRestaurant>?, userInput: String?) {
        if (userList == null || lunchRestaurantList == null) {
            return
        }

        val recyclerViewItemViewStateList: List<WorkmatesItemViewState> = lunchRestaurantList.mapNotNull { lunchRestaurant ->
            val matchingUser = userList.find { it.id.equals(lunchRestaurant.userId, ignoreCase = true) } ?: return@mapNotNull null

            WorkmatesItemViewState(
                matchingUser.photoUrl,
                matchingUser.name,
                lunchRestaurant.restaurantName,
                lunchRestaurant.restaurantId
            )
        }


        val recyclerViewItemViewStateList: List<WorkmatesItemViewState> = ArrayList()
        val userIdsEatingToday: MutableList<String> = ArrayList()
        if (lunchRestaurantList != null) {
            for (lunchRestaurant in lunchRestaurantList) {
                var matchingUser: User? = null
                for (user in userList) {
                    if (user.id.equals(lunchRestaurant.userId, ignoreCase = true)) {
                        matchingUser = user
                        break
                    }
                }
                if (matchingUser != null) {
                    userIdsEatingToday.add(matchingUser.id)
                    if (userInput == null || matchingUser.name.contains(userInput) || lunchRestaurant.restaurantName.contains(userInput)) {
                        recyclerViewItemViewStateList.add(
                            WorkmatesItemViewState(
                                matchingUser.photoUrl,
                                matchingUser.name,
                                lunchRestaurant.restaurantName,
                                lunchRestaurant.restaurantId
                            )
                        )
                    }
                }
            }
        }
        for (user in userList) {
            if (!userIdsEatingToday.contains(user.id)) {
                if (userInput == null || user.name.contains(userInput)) {
                    recyclerViewItemViewStateList.add(
                        WorkmatesItemViewState(
                            user.photoUrl,
                            user.name,
                            context.getString(R.string.not_going_to_restaurant_yet),
                            null
                        )
                    )
                }
            }
        }
        workMatesFragmentViewStateMediatorLiveData.value = WorkmatesViewState(recyclerViewItemViewStateList)
    }

    val workmatesFragmentViewStateLiveData: LiveData<WorkmatesViewState>
        get() = workMatesFragmentViewStateMediatorLiveData
}